openjdk/mmtkHeap.hpp

```diff
@@ -181,9 +181,11 @@ class MMTkHeap : public CollectedHeap {
 
   // An object is scavengable if its location may move during a scavenge.
   // (A scavenge is a GC which is not a full GC.)
-  bool is_scavengable(oop obj);
+  inline bool is_scavengable(oop obj) { return true; }
```

Comment:
This is only true for generational GCs right? Other full heap GCs like `MarkCompact` won't be scavenging

Reply:
I'm not sure as well. But from my experience, any moving GC in mmtk should return true here, otherwise it may crash for some reason...

Reply:
Seems like ZGC returns `false` [1] while Shenandoah returns `true` [2]. Other GCs will conditionally return `true` if the object is in the nursery. Maybe we should expose a `is_movable_in_current_gc()` from mmtk-core that answers this question instead of unconditionally returning `true`.

Unsure why ZGC returns `false` though. I guess it technically isn't generational. But then why does Shenandoah return `true` without being generational. Does Shenandoah do incremental collections?

\[1]: https://github.com/mmtk/openjdk/blob/jdk-11.0.15%2B8-mmtk/src/hotspot/share/gc/z/zCollectedHeap.cpp#L109-L111
\[2]: https://github.com/mmtk/openjdk/blob/jdk-11.0.15%2B8-mmtk/src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp#L509-L510

Reply:
I think normally if you state a GC will move objects but actually it won't move objects, it is more likely to be a performance issue rather than a correctness issue. If a GC is not moving, the runtime can make some assumptions about it to improve performance.

But it does not seem to be the case: before the change, we tell the runtime that we won't move objects in a 'scavenge GC', but we actually move objects in quite a few GC plans. I am wondering if this function is actually used or not. If we does not even tell the runtime that we are doing 'scavenge GC', then this function is probably irrelevant.

I am wondering what is the reason for changing the function.

Reply:
It seems like the `is_scavengable()` function is primarily used to check if an oop can be embedded in code (if it is scavengable, then it can't be embedded in code) and for the `do_oop()` visitor for the `DetectScavengableRoot` oop closure for nmethods.

Reply:
Agreed. However, the change here is simply adding an `inline` attribute so that the function body can be inlined. We can switch to a more correct implementation in the future.